### PR TITLE
refactor: use macro to iterate fields in struct or elements in list

### PR DESCRIPTION
**Pull request by neverchanje**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Formerly, to iterate the fields in a StructRef, we had to first collect the fields in a Vec. The creation of the Vec happens in the critical path of struct, such as Serialize, Display, etc. As an optimization, now I use the iterator macros, `iter_fields_ref ` and `iter_fields`, to reduce the unnecessary overhead. Likewise, I did this refactoring on List.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/common/src/array/list_array.rs [hunk 1]
```diff
@@ -382,11 +389,40 @@ pub enum ListRef<'a> {
...
+            }
+        }
+    };
+}
+
+macro_rules! iter_elems {
+    ($self:ident, $it:ident, { $($body:tt)* }) => {
+        match $self {
+            ListRef::Indexed { arr, idx } => {
+                let $it = (arr.offsets[*idx]..arr.offsets[*idx + 1]).map(|o| arr.value.value_at(o).clone().to_owned_datum());
```
Comment (wzzzzd) [inline src/common/src/array/list_array.rs:411]: Is `clone()` necessary here?

File: src/common/src/array/struct_array.rs [hunk 2]
```diff
@@ -370,54 +389,67 @@ pub enum StructRef<'a> {
...
-                                Some(s) => s.into_scalar_impl().to_protobuf(),
-                            }
-                        })
-                        .collect_vec(),
-                };
-                value.encode_to_vec()
+                let $it = arr
+                    .children
+                    .iter()
+                    .map(|a| a.value_at(*idx).clone().to_owned_datum());
```
Comment (wzzzzd) [inline src/common/src/array/struct_array.rs:414]: same

File: src/common/src/array/list_array.rs (hunk 1 above)
Comment (neverchanje) [reply src/common/src/array/list_array.rs]: No, seems clippy didn't check here. 👍

**Review by wzzzzd — APPROVED** [review]
LGTM

**Pull request merged**